### PR TITLE
AS-632: Add enableRequesterPays

### PR DESCRIPTION
**Pull request by ansingh7115**
**PR checklist**
- [ ] For each library you've modified here, decide whether it requires a major, minor, or no version bump. (Click [here](/broadinstitute/workbench-libs/blob/develop/CONTRIBUTING.md) for guidance)

If you're doing a **major** or **minor** version bump to any libraries:
- [ ] Bump the version in `project/Settings.scala` `createVersion()`
- [ ] Update `CHANGELOG.md` for those libraries
- [ ] I promise I used `@deprecated` instead of deleting code where possible

In all cases:
- [ ] Replace the appropriate version hashes in `README.md` and the `CHANGELOG.md` for any libs you changed with `TRAVIS-REPLACE-ME` to auto-update the version string
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green (CI _and_ coverage tests)
- [ ] Squash commits and **merge to develop**
- [ ] Delete branch after merge

### CODE REVIEW

**Review by MatthewBemis — COMMENTED** [review]
The change looks good. Can you also add the changes to the README and serviceTest CHANGELOG to note what updated and do the `TRAVIS-REPLACE-ME` dance?

**Comment by davidangb** [comment]
+1 to Matt's comment; the code change itself looks good.

**Comment by ansingh7115** [comment]
Whoops! Totally forgot...

**Review by MatthewBemis — APPROVED** [review]

**Review by davidangb — APPROVED** [review]

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/dao/Google.scala [hunk 1]
```diff
@@ -17,18 +19,30 @@ object Google {
...
+  lazy val googleStorageDAO =
+    new HttpGoogleStorageDAO(appName, pemMode, metricBaseName)(system, ec)
   lazy val googleDirectoryDAO =
     new HttpGoogleDirectoryDAO(appName, pemModeWithServiceAccountUser, metricBaseName)(system, ec)
+
+  def storageOptions(projectName: String, googleCredentials: GoogleCredentials): Storage = StorageOptions.newBuilder
+    .setProjectId(projectName)
+    .setCredentials(googleCredentials)
+    .build
+    .getService
```
Comment (ansingh7115) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/dao/Google.scala]: this `storageOptions` is the only change in this file, everything else is auto formatting

Comment (davidangb) [reply serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/dao/Google.scala]: I think this change breaks the encapsulation model between `serviceTest` and `google`. To follow the current pattern, I think you should:
* add a new `setRequesterPays` method to `o.b.d.workbench.google.HttpGoogleStorageDAO`, and undo all changes in `serviceTest`
* import the newer snapshot of `google` in Rawls
* in Rawls' automation `WorkspaceApiSpec`, `import o.b.d.workbench.dao.Google. googleStorageDAO` and simply call `googleStorageDAO.setRequesterPays()`. See `RawlsApiSpec` for example usages.

**Review by davidangb — CHANGES_REQUESTED** [review]

File: project/Dependencies.scala [hunk 2]
```diff
@@ -207,7 +207,8 @@ object Dependencies {
     jacksonModule,
     akkaStream,
     rawlsModel,
-    scalaTestSelenium
+    scalaTestSelenium,
+    googleStorageNew
```
Comment (davidangb) [inline project/Dependencies.scala]: can `googleStorageNew` now be removed from serviceTest dependencies?

File: google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleStorageDAO.scala [hunk 3]
```diff
@@ -59,6 +59,51 @@ class HttpGoogleStorageDAO(appName: String,
...
+                        Option(entity),
+                        endTime - startTime,
+                        Option(response.status.intValue),
+                        None
+          ).toJson(GoogleRequestFormat).compactPrint
+        )
+        ()
+      }
+    }
+  }
```
Comment (davidangb) [inline google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleStorageDAO.scala]: I am ok with this approach if it's the only feasible one, but if the client libraries work, I think we should try to use those. Something like this, maybe? This is untested, I don't know if it has drawbacks:
```
    import com.google.api.services.storage.model.{Bucket=>GoogleBucket}

    val rp = new Bucket.Billing().setRequesterPays(true)
    storage.buckets().patch(bucketName, new GoogleBucket().setBilling(rp))
```

File: google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleStorageDAO.scala [hunk 4]
```diff
@@ -59,6 +59,51 @@ class HttpGoogleStorageDAO(appName: String,
   private lazy val storage =
     new Storage.Builder(httpTransport, jsonFactory, googleCredential).setApplicationName(appName).build()
 
+  def enableRequesterPays(projectName: String, bucketName: String): Future[Unit] = {
```
Comment (davidangb) [inline google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleStorageDAO.scala]: how about making this `setRequesterPays` and accept a Boolean, so we can turn off RP as well?

File: google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleStorageDAO.scala (hunk 3 above)
Comment (davidangb) [reply google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleStorageDAO.scala]: might also want to wrap this inside a call to `getBucket` - see if the current RP value is already what we want it to be; if it is, skip the patch call entirely

Comment (ansingh7115) [reply google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleStorageDAO.scala]: Thank you - I didn't realize you could just use Billing() - I was following https://cloud.google.com/storage/docs/using-requester-pays#code-samples

File: README.md [hunk 5]
```diff
@@ -106,7 +106,7 @@ Workbench utility libraries, built for Scala 2.12 and 2.13. You can find the ful
 
 Contains common classes and utilities for writing tests against Workbench REST web services.
 
-Latest SBT dependency: `"org.broadinstitute.dsde.workbench" %% "workbench-service-test" % "0.18-23f5ae7" % "test" classifier "tests"`
+Latest SBT dependency: `"org.broadinstitute.dsde.workbench" %% "workbench-service-test" % "0.18-TRAVIS-REPLACE-ME" % "test" classifier "tests"`
```
Comment (MatthewBemis) [inline README.md]: this change should be in the `google` section, right?

**Review by MatthewBemis — APPROVED** [review]
👍 pending the README.md fix

File: README.md (hunk 5 above)
Comment (ansingh7115) [reply README.md]: ugh why am i so forgetful

**Review by davidangb — APPROVED** [review]
:+1: code looks good! Thanks for all the iteration on this one.

Last thing: have you - and if not, could you - just try the patch against a real bucket and look at the debug logs to see what's being sent, and ensure it has the right effect on the bucket?

**Comment by ansingh7115** [comment]
I made the test log the bucket details before and after enabling RP: 

Bucket Details before enabling:

```
{
  "etag": "CAI=",
  "iamConfiguration": {
    "bucketPolicyOnly": {
      "enabled": true,
      "lockedTime": "2021-05-17T04:12:57.013Z"
    },
    "uniformBucketLevelAccess": {
      "enabled": true,
      "lockedTime": "2021-05-17T04:12:57.013Z"
    }
  },
  "id": "fc-831b1be9-3689-474e-a890-b1594bdaf5f2",
  "kind": "storage#bucket",
  "labels": {
    "security": "low"
  },
  "location": "US",
  "locationType": "multi-region",
  "logging": {
    "logBucket": "storage-logs-gpalloc-dev-master-yruq8mo",
    "logObjectPrefix": "fc-831b1be9-3689-474e-a890-b1594bdaf5f2"
  },
  "metageneration": "2",
  "name": "fc-831b1be9-3689-474e-a890-b1594bdaf5f2",
  "projectNumber": "104034583618",
  "selfLink": "https://www.googleapis.com/storage/v1/b/fc-831b1be9-3689-474e-a890-b1594bdaf5f2",
  "storageClass": "STANDARD",
  "timeCreated": "2021-02-16T04:12:57.013Z",
  "updated": "2021-02-16T04:12:57.850Z"
}
```
Bucket Details after enabling:
```
{
  "billing": {
    "requesterPays": true
  },
  "etag": "CAM=",
  "iamConfiguration": {
    "bucketPolicyOnly": {
      "enabled": true,
      "lockedTime": "2021-05-17T04:12:57.013Z"
    },
    "uniformBucketLevelAccess": {
      "enabled": true,
      "lockedTime": "2021-05-17T04:12:57.013Z"
    }
  },
  "id": "fc-831b1be9-3689-474e-a890-b1594bdaf5f2",
  "kind": "storage#bucket",
  "labels": {
    "security": "low"
  },
  "location": "US",
  "locationType": "multi-region",
  "logging": {
    "logBucket": "storage-logs-gpalloc-dev-master-yruq8mo",
    "logObjectPrefix": "fc-831b1be9-3689-474e-a890-b1594bdaf5f2"
  },
  "metageneration": "3",
  "name": "fc-831b1be9-3689-474e-a890-b1594bdaf5f2",
  "projectNumber": "104034583618",
  "selfLink": "https://www.googleapis.com/storage/v1/b/fc-831b1be9-3689-474e-a890-b1594bdaf5f2",
  "storageClass": "STANDARD",
  "timeCreated": "2021-02-16T04:12:57.013Z",
  "updated": "2021-02-16T04:12:59.999Z"
}
```

It appears that the details prior to enabling RP don't have `"billing"`

**Pull request merged**